### PR TITLE
Spectromeasurements

### PR DESCRIPTION
**Pull request by ScottSoren**
This solves all of the open issues labeled "spectra", plus a few others, and gets all tests passing as intended.

Most of the code changes were associated with #158.

#165 is already incorporated (thanks @matenestor !)

@AnnaWiniwarter , if you have any other datasets to test this on, might catch a residual bug sooner rather than later :)

### CODE REVIEW

File: CHANGES.rst
Comment (AnnaWiniwarter) [inline CHANGES.rst:43]: I'm confused, I just tested it and had to actively include it (i.e. pass` include_spectra=True`) in order for it to read the spectra and return a `SpectroMSMeasurement`. 

Also, here you call it `SpectroMSMeasurement` but a few lines down `MSSpectroMeasurement` - which one is it now?

**Review by AnnaWiniwarter — COMMENTED** [review]
I don't have time right now for a full code review (and Jakub is better at that anyway), but I tested it on some different combinations of data I have lying around (did not test it for the dataset I sent you, @ScottSoren which is the only one I have with EC, MS and spectra in one set). 
When trying to add two `MSSpectroMeasurement`s, I get this (for me) unexpected error.
image

The measurements have a different number of spectra and (more likely the cause of the error) different settings for the spectra, i.e. a range of 0-50 amu for one and 0-100 for the other one.
Just to be sure you're aware.

File: CHANGES.rst
Comment (ScottSoren) [reply CHANGES.rst:43]: It's now `MSSpectroMeasurement` , will fix!
I noticed that but thought I had fixed it so that it would automatically read spectra again! (It does now when there's also EC in the dataset.) Will fix.

**Comment by ScottSoren** [comment]
@AnnaWiniwarter , in this case I must say the error is expected for me. If it should do something other than raise an error when appending measurements with spectra of different shapes, we have to decide what. What do you think it should do?

**Comment by AnnaWiniwarter** [comment]
I wonder if you could make the error message more explicit for this case? I.e. specify that it can't append to `MSSpectroMeasurements` if the mass range (does it also apply to a different mass resolution?) is different. Would that be possible?

**Comment by matenestor** [comment]
I am in middle of the review. Should be done with it tomorrow, Thursday latest.

Regarding the appending error. I also think it could be more explicit, if it really is not possible to add them together, because of logical scientific reasons. Although, they are not clear to me right now. Why different amount of spectra couldn't be added together?
- If they have the same AMU range, is it then like doing "2 + 3 = 5"? Meaning a list with 2 spectra plus a list with 3 spectra.
- If they don't, then plotting them together with MS plot should be fine, since the spectra match the breaks in the MS plot. So what is preventing it?

**Comment by ScottSoren** [comment]
@matenestor , 
Great! I will wait on your review :) 
To answer, in the two cases you present:
- Appending `spectrum_series` with different numbers of spectra should be no problem, like you said. If an error came up then, it would definitely be a bug to fix.
- If the AMU range is different, you get the error Anna describes. Good point that it should be possible to plot in a sensible way. The problem is the `spectrum_series` data structure - this is not just a list of `spectrum`s, but a 2-D data series. It can only represent spectra with the same number of AMU points because a numpy array has to be rectangular. Which leaves the following options. (Keep in mind that a lot of analysis tools are likely to come on top of the data structure that we choose.)
  - ditch `SpectrumSeries` and just allow an `MSMeasurement` to have a list of associated `Spectrum` objects. Then each spectrum object can have its own x series (AMU range), so you would gain flexibility on that front. What you loose is the ability to do vectorized analysis of the spectra - that might not matter for the cases so far, but if someone is taking continuous spectra overnight, they would feel the difference between treating the data as a 2-D array vs looping over 1-D arrays. For other techniques with faster spectrum acquisition times, like UV-Vis, this difference is more critical.
  - allow for lists of `SpectrumSeries` when the x series is not constant. The problem here is that it makes all the methods acting on the spectra much less intuitive - you'd always have to specify which `SpectrumSeries` you are acting on.
  - do some kind of manipulation when appending `SpectrumSeries` objects to get them on the same `xseries`. That could be interpolating the spectra with the larger x range onto the smaller `xseries` or filling the smaller spectra with 0's or None's to cover the larger `xseries`. The disadvantage is that it might not be intuitive or transparent, and you wouldn't get the plot that you describe with the AMU range shown clearly in the range of the spectra.
  -- Raise an error message when `SpectrumSeries` with different `xseries` are added together, but point the user towards a method (like `cast_onto_x`) that can be used on a `SpectrumSeries` or `SpectroMeasurement` to explicitly change the x range of the spectra. This method could take inputs allowing for the above options.

It may be tied into [Kenneth's comment here](https://github.com/ixdat/ixdat/issues/158#issuecomment-1934870286).

Let me know what you think. And I look forward to your review! If you can catch why available spectra are not always loaded by default, as Anna noticed, that's for sure something I hope to fix.

File: src/ixdat/exporters/ms_exporter.py
Comment (matenestor) [inline src/ixdat/exporters/ms_exporter.py:16]: Is it possible that the `SpectrumSeriesExporter` would take the whole `measurement` only like this exporter is doing? I sense an inconsistency

**Review by matenestor — CHANGES_REQUESTED** [review]
I have doubts whether combining multiple already specialized classes to create new ones like `MSSpectroMeasurement` and `ECMSSpectroMeasurement` etc. is a good idea. We are thinking with @KennethNielsen about an alternative way this codebase could go. We will make a proposal in the future.
Since this is a PR to fix stuff, let's keep it like this. It won't affect users anyway. But us developers, we will benefit.

Now I am thinking whether having a "technique" with `-MS_spectra` suffix is also relevant. Offering `include_mass_scans` only could make the interface more friendly. There is a comment about that.

Other than that, I have a few questions.

**Comment by matenestor** [comment]
I am still getting an error when combining an EC measurement with spectra. I don't think I have found your suggestion, but what it should call would probably be `ECMeasurement` class and then it would combine it with `SpectroMeasurement`? That really smells like another `ECSpectroMeasurement` class, though :smile:

File: src/ixdat/readers/zilien.py
Comment (ScottSoren) [reply src/ixdat/readers/zilien.py:1]: A discussion of how zilien names and organizes files is needed but out of scope here ;)

File: src/ixdat/plotters/ms_plotter.py
Comment (ScottSoren) [reply src/ixdat/plotters/ms_plotter.py:1]: Good point. This is pervasive in `plotters` as well as `exporters`. I'll fix it.

File: src/ixdat/measurements.py
Comment (ScottSoren) [reply src/ixdat/measurements.py:231]: Nice! Thanks, that is an improvement.

File: src/ixdat/spectra.py
Comment (ScottSoren) [reply src/ixdat/spectra.py:819]: I don't understand. What should `SpectroMeasurement.join()` do?

File: src/ixdat/exporters/spectrum_exporter.py
Comment (ScottSoren) [reply src/ixdat/exporters/spectrum_exporter.py:20]: Good point. But it was obviously wrong before! The reference `spectrum` is given to the exporter during its initialization, whereas `path_to_file` should be specified each time it makes an export.
I don't think it was in direct use by anyone: Neither `Spectrum.export` nor `SpectrumSeries.export` worked before (#139). `SECMeasurement.export()` did work, and is updated to continue to work without API change.
If I'm wrong, I think we hope the user can fix it: The deprecation seems a bit messy to implement here, and we're not strictly committed to backwards compatibility before v1.0.0.

File: src/ixdat/exporters/spectrum_exporter.py
Comment (ScottSoren) [reply src/ixdat/exporters/spectrum_exporter.py:69]: Same answer :)

File: src/ixdat/techniques/ms.py
Comment (ScottSoren) [reply src/ixdat/techniques/ms.py:52]: Now I want to change it to `Calculator`s :P 
See #164

File: src/ixdat/spectra.py
Comment (ScottSoren) [reply src/ixdat/spectra.py:1]: I have put it in `Measurement.__getitem__`. It makes most sense there, no?

File: src/ixdat/techniques/ec_ms.py
Comment (ScottSoren) [reply src/ixdat/techniques/ec_ms.py:475]: It's not that wild given that we have method resolution order.

```
from ixdat.techniques import ECMSSpectroMeasurement

ECMSSpectroMeasurement.mro()
Out[121]: 
[ixdat.techniques.ec_ms.ECMSSpectroMeasurement,
 ixdat.techniques.ec_ms.ECMSMeasurement,
 ixdat.techniques.ec.ECMeasurement,
 ixdat.techniques.ms.MSSpectroMeasurement,
 ixdat.techniques.ms.MSMeasurement,
 ixdat.spectra.SpectroMeasurement,
 ixdat.measurements.Measurement,
 ixdat.db.Saveable,
 object]
 ```
 Looks good to me!

File: src/ixdat/techniques/ms.py
Comment (ScottSoren) [reply src/ixdat/techniques/ms.py:1409]: Yes.
This particular decorator just ensures that the `siq_quantifier` from the original measurement is attached to the cut measurement as well. Other stuff is taken care of for this purposes by `PlaceHolderObject`s and `MemoryBackend`. I must admit I still don't know how to have this kind of persistence representation for external calculators, even assuming we implement #164.

File: src/ixdat/readers/zilien.py
Comment (ScottSoren) [reply src/ixdat/readers/zilien.py:120]: This is in the Zilien reader so I will leave perfecting it to you! I have put in a link to this comment.

File: src/ixdat/spectra.py
Comment (ScottSoren) [reply src/ixdat/spectra.py:723]: Thanks for catching this! It has apparently not been tested. I've added the parentheses.

File: src/ixdat/spectra.py
Comment (ScottSoren) [reply src/ixdat/spectra.py:845]: Sort of. #75 . And/or `sqlalchemy` integration.

**Comment by ScottSoren** [comment]
Hi @matenestor , I have implemented your review! :) 
I didn't do anything, though, on the issue of appending spectra of different m/z-spans. I think we have to deal with that seperately, becase we (I at least) don't know what exactly ixdat should do in that case (and I think the existing error message is clear enough).
I also didn't do anything about your last comment... you want to combine an `ECMeasurement` with a `MSSpectrumSeries` without any MID data? That might be do-able with just an extra line in `TECHNIQUE_CLASSES` in `techniques.__init__`: 
```
"EC-MS_spectra": SpectroECMeasurement
```
Can you give it a try? :)

File: src/ixdat/techniques/ms.py
Comment (matenestor) [reply src/ixdat/techniques/ms.py:1399]: Also no. I am getting this error when trying to print `measurement.t`

```python
  File "/home/jkm/ws/ecms/ixdat/src/ixdat/measurements.py", line 914, in t
    return self[self.control_series_name].t
  File "/home/jkm/ws/ecms/ixdat/src/ixdat/measurements.py", line 705, in __getitem__
    series = self.get_series(key)
  File "/home/jkm/ws/ecms/ixdat/src/ixdat/measurements.py", line 778, in get_series
    if key.endswith("-t") or key.endswith("-x"):
AttributeError: 'NoneType' object has no attribute 'endswith'
```

Comment (matenestor) [reply src/ixdat/techniques/ms.py:1399]: One more thing. In shell, when I have `ECMSMeasurement` and I type a dot for autocompletion (Python then executes some of ixdat's code in the background) I am getting the warnings from `‎Measurement._build_selector_series()`. I don't know if it happens with every `ECMSMeasurement` though. I using those in `ixdat-large-test-files`

image

File: src/ixdat/spectra.py
Comment (matenestor) [reply src/ixdat/spectra.py:822]: Aha, that is a consequence of inheriting from `Measurement` class. Since the `join()` method there raises `NotImplementedError` it means it is abstract, and it should be implemented in all classes that inherit from `Measurement`, otherwise the exception is raised. 
And when `SpectroMeasurement` is a measurement with spectrum series attached, I assumed it will be possible to join it with other `SpectroMeasurement`s?

File: development_scripts/reader_demonstrators/demo_zilien_spectrum_reader.py
Comment (matenestor) [reply development_scripts/reader_demonstrators/demo_zilien_spectrum_reader.py:92]: `MSMeasurement` has a `t` variable (property), it inherits it from `Measurement`. Does it make sense for `MSMeasurement` to have some kind of `control_series_name`? The `t` is taken from there

---

I reread the other comment where you write

> what the primary time series should be in an MS measurement, where each valueseries has its own timeseries.
> I suggest leaving this for a later PR

Agreed

File: src/ixdat/exporters/ms_exporter.py
Comment (matenestor) [reply src/ixdat/exporters/ms_exporter.py:33]: Looks like a typo and should say "see CSVExporter.prepare_header_and_data"

File: src/ixdat/spectra.py
Comment (matenestor) [reply src/ixdat/spectra.py:822]: (note: we agreed that `join()` will be removed)

**Comment by matenestor** [comment]
@ScottSoren I replied to your comments and I consider them mostly resolved. Some are left open for you to read, so know about it.

And the rest is waiting for your answer (those are the left out comments from the first round). There is three of them. Once you comment, I will approve the PR.

File: src/ixdat/exporters/csv_exporter.py
Comment (ScottSoren) [reply src/ixdat/exporters/csv_exporter.py:18]: Good point, for another PR, though, I would say. I will leave a comment here.

File: src/ixdat/techniques/ms.py
Comment (ScottSoren) [reply src/ixdat/techniques/ms.py:1399]: Huh. Interesting. I think these are separate issues, though. 

The error message from `Measurement.__getitem__` when that comes from looking up `control_series_name` = `None` should improve a bit due to the new [step 0](https://github.com/ixdat/ixdat/blob/99c60a2c8933c8e3bb6c4e1ff8db7ff5f893fe39/src/ixdat/measurements.py#L704) which checks the type of the argument. But I guess it is still confusing.
 
I've made an issue about control series linking here.

File: src/ixdat/db.py
Comment (ScottSoren) [reply src/ixdat/db.py:435]: `identity` is an `int` if the backend is the active backend. Otherwise, the backend has to be specified to fully identify the object, and so `identity` becomes a tuple. I agree it is not ideal. Let's fix it elsewhere (#75).

File: src/ixdat/readers/zilien.py
Comment (ScottSoren) [reply src/ixdat/readers/zilien.py]: I've put a comment in the code linking to this conversation.

**Comment by ScottSoren** [comment]
@matenestor , I've commented on the remaining issues!
@AnnaWiniwarter , I've made that error message more explicit.
Forgive the long period of inactivity. Hopefully it is good to merge now :)

**Pull request merged**
